packages/core/src/indexing/client.ts

```diff
@@ -0,0 +1,998 @@
+    }
+
+    // required block actions
+
+    for (const action of [
+      "getBlock",
+      "getBlockTransactionCount",
+      "getTransactionCount",
+    ] as const) {
+      // @ts-ignore
+      actions[action] = _publicActions[action];
+    }
+
+    const actionsWithMetrics = {} as PonderActions;
+
+    for (const [action, actionFn] of Object.entries(actions)) {
+      // @ts-ignore
+      actionsWithMetrics[action] = async (
+        ...args: Parameters<PonderActions[keyof PonderActions]>
+      ) => {
+        const endClock = startClock();
+        try {
+          // @ts-ignore
+          return await actionFn(...args);
+        } finally {
+          common.metrics.ponder_indexing_rpc_action_duration.observe(
+            { action },
+            endClock(),
+          );
+        }
+      };
+    }
+
+    return actionsWithMetrics;
+  };
+
+  return {
+    getClient(network) {
+      const requestQueue =
+        requestQueues[indexingBuild.networks.findIndex((n) => n === network)]!;
+
+      return createClient({
+        transport: cachedTransport({
+          common,
+          network,
+          requestQueue,
+          syncStore,
+          cache,
+        }),
+        chain: network.chain,
+        // @ts-expect-error overriding `readContract` is not supported by viem
+      }).extend(ponderActions);
+    },
+    async prefetch({ events, eventCount }) {
+      // Use profiling metadata + next event batch to determine which
+      // rpc requests are going to be made, and preload them into the cache.
+
+      const prediction: { ev: number; request: Request }[] = [];
+
+      for (const event of events) {
+        if (profile.has(event.name)) {
+          for (const [, { pattern, count }] of profile.get(event.name)!) {
+            // Expected value of times the prediction will be used.
+            const ev = count / eventCount[event.name]!;
+            prediction.push({
+              ev,
+              request: recoverProfilePattern(pattern, event),
+            });
+          }
+        }
+      }
+
+      const chainRequests: Map<
+        number,
+        { ev: number; request: EIP1193Parameters }[]
+      > = new Map();
+      for (const network of indexingBuild.networks) {
+        chainRequests.set(network.chainId, []);
+      }
+
+      for (const { ev, request } of dedupe(prediction, ({ request }) =>
+        getCacheKey(encodeRequest(request)),
+      )) {
+        chainRequests.get(request.chainId)!.push({
+          ev,
+          request: encodeRequest(request),
+        });
+      }
+
+      for (const [chainId, requests] of chainRequests.entries()) {
+        const ni = indexingBuild.networks.findIndex(
+          (n) => n.chainId === chainId,
+        );
+        const network = indexingBuild.networks[ni]!;
+        const requestQueue = requestQueues[ni]!;
+
+        const dbRequests = requests.filter(({ ev }) => ev > 0.2);
+
+        const cachedResults = await syncStore.getRpcRequestResults({
```

Comment:
Maybe worth running concurrently across chains? Or even batching into one db query across all chains?

Reply:
Gonna just do await Promise.all for now